File: Irrgarten_ruby/combat_element.rb
# encoding:utf-8
require_relative 'dice'
module Irrgarten
class CombatElement
    # Initializes a new CombatElement object with the given effect and uses.
    def initialize(effect,uses)
        @effect = effect
        @uses = uses
    end

    # @return the effect of the CombatElement
    def produce_effect
        if @uses > 0
            @uses -= 1
            @effect
        else
            0.0
        end
    end

    # Discards the CombatElement from the player's inventory
    # @return [bool] true if the element was discarded, false otherwise
    def discard
        Dice.discard_element(@uses)
    end

    # @return [String] a string representation of the CombatElement
    def to_s
        "[#{@effect},#{@uses}]"
    end

    protected :produce_effect
end

end

File: Irrgarten/src/irrgarten/CombatElement.java
package irrgarten;

/**
 * Abstract class for a combat element. A combat element is an element that can
 * be used in combat to affect the outcome of the combat.
 */
public abstract class CombatElement {
    /**
     * The effect of the combat element.
     */
    private float effect;

    /**
     * The number of uses of the combat element.
     */
    private int uses;
    
    /**
     * Constructor for the combat element. Initializes the combat element with the
     * effect and the number of uses.
     * @param effect The effect of the combat element.
     * @param uses The number of uses of the combat element.
     */
    public CombatElement(float effect, int uses){
        this.effect = effect;
        this.uses = uses;
    }
    
    /**
     * Produces the effect of the combat element. If the combat element has no
     * uses left, the effect is 0.
     * @return The effect of the combat element.
     */
    protected float produceEffect(){
        float result;
        if (this.uses > 0){
            result = this.effect;
            this.uses--;
        }    
        else
            result = 0f;
        return result;
    }
    
    /**
     * Decides if a combat element must be discarded
     * 
     * @return true if the combat element must be discarded, false otherwise
     */
    public boolean discard(){
        return Dice.discardElement(this.uses); 
    }
    
    /**
     * Return a string representation of the combat element.
     * @return A string representation of the combat element.
     */
    public String toString(){
        return "[" + this.effect + "," + this.uses + "]";
    };
}
